### ISSUE TITLE
A Recipe for Training Neural Networks

### ISSUE THREAD

**Issue by amyxie361** [amyxie361/Blog#4]
A Recipe for Training Neural Networks
https://karpathy.github.io/2019/04/25/recipe/
A good training guide for NN from Victor

**1. Become one with the data**
read raw data:
- corrupted data
- distribution, unbalanced, bias
- how human do prediction: what feature are we lookin at
- possible preprocess
- noise?

read prediction:
- find possible error

filter out certain kind of data

**2. Set up training/evaluation skeleton + baselines**
set up training/eval process use a simple baseline
- fix random seed.
- simplify: switch off all parameters
- add significant digits to your eval: plot the loss, with digits
- verify loss @ init: estimate the loss with random init
- init well: init final layer with the same as distribution will help speed up.
- human baseline: monitor with acc
- input-independent baseline: model should be better with a input
- overfit one batch: get the lowest loss, and check the label and the prediction
- verify decreasing training loss: increase capacity, the loss should go down.
- visualize just before the net: immediately before your y_hat = model(x) (or sess.run in tf). visualize exactly what goes into your network
- visualize prediction dynamics: visualize model predictions on a fixed test batch. 
Many times it is possible to feel the network “struggle” to fit your data if it wiggles too much in some way, revealing instabilities. Very low or very high learning rates are also easily noticeable in the amount of jitter.
- use backprop to chart dependencies ???
- generalize a special case: write a toy case from scratch.

**3. overfit**
first get a model large enough that it can overfit (i.e. focus on training loss) and then regularize it appropriately (give up some training loss to improve the validation loss). 

- picking the model.
find the most related paper and copy-paste their simplest architecture that achieves good performance
- adam is safe.
- complexify only one at a time: every time ensure that you get a performance boost you’d expect
- do not trust learning rate decay defaults: use fixed lr and tune by hand

**4. Regularize**
- get more data: add more real data, ensembles(5 models)
- data augment
- creative augmentation: For example, domain randomization, use of simulation, clever hybrids such as inserting (potentially simulated) data into scenes, or even GANs.
- pretrain
- stick with supervised learning
- smaller input dimensionality. Remove features that may contain spurious signal.
- smaller model size
- decrease the batch size: normalization inside batch norm
- Add dropout. dropout does not seem to play nice with batch normalization
https://arxiv.org/abs/1801.05134 Understanding the Disharmony between Dropout and Batch Normalization by Variance Shift
- weight decay. Increase the weight decay penalty.
- early stopping
- try a larger mode

**5. Tune**
- random over grid search
http://jmlr.csail.mit.edu/papers/volume13/bergstra12a/bergstra12a.pdf
Random Search for Hyper-Parameter Optimization
- hyper-parameter optimization
bayesian hyper-parameter optimization toolboxes

**6. Squeeze out the juice**
- ensembles: gain 2% of accuracy on anything
https://arxiv.org/abs/1503.02531
Distilling the Knowledge in a Neural Network
- leave it training

**Comment by amyxie361** [comment]
A Survival Guide to a PhD
https://karpathy.github.io/2016/09/07/phd/

http://paperscape.org/
https://www.quora.com/I-got-a-job-offer-from-Google-Facebook-and-Microsoft-and-I-also-got-accepted-into-the-Ph-D-in-Computer-Science-programs-at-MIT-Stanford-and-Berkeley-What-factors-should-I-consider-while-making-a-choice-between-the
